Regression: ddev list, with appdir deleted, gives a panic

## What happened (or feature request):

In https://github.com/drud/ddev/pull/444 we made it possible to do a `ddev list` even if an app's directory was removed.  Unfortunately there's been a regression and we're worse than we were before: Now we get a panic in this case.

```
$ ddev list
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1428df7]

goroutine 1 [running]:
github.com/drud/ddev/pkg/plugins/platform.GetApps(0x0)
	/go/src/github.com/drud/ddev/pkg/plugins/platform/utils.go:55 +0x3f7
github.com/drud/ddev/cmd/ddev/cmd.glob..func11(0x19a90a0, 0x19d57e0, 0x0, 0x0)
	/go/src/github.com/drud/ddev/cmd/ddev/cmd/list.go:17 +0x34
github.com/drud/ddev/vendor/github.com/spf13/cobra.(*Command).execute(0x19a90a0, 0x19d57e0, 0x0, 0x0, 0x19a90a0, 0x19d57e0)
	/go/src/github.com/drud/ddev/vendor/github.com/spf13/cobra/command.go:654 +0x2a2
github.com/drud/ddev/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x19a9be0, 0xc42001c0b8, 0x0, 0xc42013df48)
	/go/src/github.com/drud/ddev/vendor/github.com/spf13/cobra/command.go:729 +0x2fe
github.com/drud/ddev/vendor/github.com/spf13/cobra.(*Command).Execute(0x19a9be0, 0x0, 0x0)
	/go/src/github.com/drud/ddev/vendor/github.com/spf13/cobra/command.go:688 +0x2b
github.com/drud/ddev/cmd/ddev/cmd.Execute()
	/go/src/github.com/drud/ddev/cmd/ddev/cmd/root.go:105 +0x3b
main.main()
	/go/src/github.com/drud/ddev/cmd/ddev/main.go:6 +0x20
```

## What you expected to happen:


## How to reproduce this:

With current master, v0.9.3-3-g4ca6d38a (g4ca6d38a), 
1. start an app, 
2. remove its directory, 
3. do a `ddev list`

## Version: Please include the output of `ddev version`, `docker version` and the project's .ddev/config.yaml.

```
n$ ddev version
cli:       	v0.9.3-3-g4ca6d38a
web:       	drud/nginx-php-fpm-local:v0.8.0
db:        	drud/mysql-local-57:v0.6.2
dba:       	drud/phpmyadmin:v0.2.0
router:    	drud/ddev-router:v0.4.3
commit:    	v0.9.3-3-g4ca6d38a
build info:	Built Tue Oct 24 15:01:34 MDT 2017 rfay@Randy-Air.local drud/golang-build-container:v0.5.0
```

## Anything else do we need to know:


## Related source links or issues:

https://github.com/drud/ddev/pull/444 was the issue where we solved this, and v0.9.3 doesn't have any trouble with it, so the regression is likely introduced in https://github.com/drud/ddev/pull/495, the command-line config changes. 

We also have a pending, nearly-done and highly-related PR in https://github.com/drud/ddev/pull/460 which solves the problem for `ddev rm` and others.  

I guess we probably need a test for this case if we're going to support it.

It looks to me like the panic is in platform.GetApps, 

https://github.com/drud/ddev/blob/4ca6d38aa2158595160231fd3ac7aeec2501f3bc/pkg/plugins/platform/utils.go#L48-L56

When we try to assign to siteStruct.AppConfig.Name we get the panic.

--- a/cmd/ddev/cmd/list.go
+++ b/cmd/ddev/cmd/list.go
@@ -27,10 +27,10 @@ var DevListCmd = &cobra.Command{
 		table := platform.CreateAppTable()
 		for _, site := range sites {
 			desc, err := site.Describe()
-			appDescs = append(appDescs, desc)
 			if err != nil {
 				util.Failed("Failed to describe site %s: %v", site.GetName(), err)
 			}
+			appDescs = append(appDescs, desc)
 			platform.RenderAppRow(table, desc)
 		}
 

--- a/cmd/ddev/cmd/start.go
+++ b/cmd/ddev/cmd/start.go
@@ -33,7 +33,7 @@ provide a working environment for development.`,
 func appStart() {
 	app, err := platform.GetActiveApp("")
 	if err != nil {
-		util.Failed("Failed to start %s, err: %v", app.GetName(), err)
+		util.Failed("Failed to start: %v", err)
 	}
 
 	output.UserOut.Printf("Starting environment for %s...", app.GetName())

--- a/pkg/plugins/platform/local.go
+++ b/pkg/plugins/platform/local.go
@@ -45,9 +45,12 @@ func (l *LocalApp) GetType() string {
 // Init populates LocalApp settings based on the current working directory.
 func (l *LocalApp) Init(basePath string) error {
 	config, err := ddevapp.NewConfig(basePath, "")
+
+	// Save config to l.AppConfig so we can capture and display the site's
+	// status regardless of its validity
+	l.AppConfig = config
+
 	if err != nil {
-		// Save config to l.AppConfig so we can capture and display the site's status.
-		l.AppConfig = config
 		return err
 	}
 
@@ -55,8 +58,6 @@ func (l *LocalApp) Init(basePath string) error {
 	if err != nil {
 		return err
 	}
-
-	l.AppConfig = config
 
 	web, err := l.FindContainerByType("web")
 	if err == nil {

--- a/pkg/plugins/platform/local_test.go
+++ b/pkg/plugins/platform/local_test.go
@@ -676,9 +676,6 @@ func TestListWithoutDir(t *testing.T) {
 	// Make sure we move out of the directory for Windows' sake
 	garbageDir := testcommon.CreateTmpDir("RestingHere")
 	defer testcommon.CleanupDir(garbageDir)
-	// Changing directory must be pushed on defer stack last so it happens
-	// before any cleanups, so windows tests won't break.
-	defer os.Chdir(packageDir)
 
 	err = os.Chdir(garbageDir)
 	assert.NoError(err)
@@ -704,6 +701,11 @@ func TestListWithoutDir(t *testing.T) {
 	assert.Contains(table.String(), fmt.Sprintf("app directory missing: %s", testDir))
 
 	err = app.Down(true)
+	assert.NoError(err)
+
+	// Change back to package dir. Lots of things will have to be cleaned up
+	// in defers, and for windows we have to not be sitting in them.
+	err = os.Chdir(packageDir)
 	assert.NoError(err)
 }
 

--- a/pkg/plugins/platform/local_test.go
+++ b/pkg/plugins/platform/local_test.go
@@ -177,6 +177,27 @@ func TestLocalStart(t *testing.T) {
 	testcommon.CleanupDir(another.Dir)
 }
 
+// TestStartWithoutDdev makes sure we don't have a regression where lack of .ddev
+// causes a panic.
+func TestStartWithoutDdevConfig(t *testing.T) {
+	// Set up tests and give ourselves a working directory.
+	assert := asrt.New(t)
+	testDir := testcommon.CreateTmpDir("TestStartWithoutDdevConfig")
+
+	// testcommon.Chdir()() and CleanupDir() check their own errors (and exit)
+	defer testcommon.CleanupDir(testDir)
+	defer testcommon.Chdir(testDir)()
+
+	err := os.MkdirAll(testDir+"/sites/default", 0777)
+	assert.NoError(err)
+	err = os.Chdir(testDir)
+	assert.NoError(err)
+
+	_, err = platform.GetActiveApp("")
+	assert.Error(err)
+	assert.Contains(err.Error(), "unable to determine")
+}
+
 // TestGetApps tests the GetApps function to ensure it accurately returns a list of running applications.
 func TestGetApps(t *testing.T) {
 	assert := asrt.New(t)

--- a/pkg/plugins/platform/local_test.go
+++ b/pkg/plugins/platform/local_test.go
@@ -646,14 +646,14 @@ func TestListWithoutDir(t *testing.T) {
 	// Set up tests and give ourselves a working directory.
 	assert := asrt.New(t)
 	testcommon.ClearDockerEnv()
+	packageDir, _ := os.Getwd()
 
 	// startCount is the count of apps at the start of this adventure
 	apps := platform.GetApps()
 	startCount := len(apps["local"])
 
 	testDir := testcommon.CreateTmpDir("TestStartWithoutDdevConfig")
-
-	defer testcommon.Chdir(testDir)()
+	defer testcommon.CleanupDir(testDir)
 
 	err := os.MkdirAll(testDir+"/sites/default", 0777)
 	assert.NoError(err)
@@ -676,6 +676,9 @@ func TestListWithoutDir(t *testing.T) {
 	// Make sure we move out of the directory for Windows' sake
 	garbageDir := testcommon.CreateTmpDir("RestingHere")
 	defer testcommon.CleanupDir(garbageDir)
+	// Changing directory must be pushed on defer stack last so it happens
+	// before any cleanups, so windows tests won't break.
+	defer os.Chdir(packageDir)
 
 	err = os.Chdir(garbageDir)
 	assert.NoError(err)
@@ -686,9 +689,17 @@ func TestListWithoutDir(t *testing.T) {
 
 	assert.EqualValues(len(apps["local"]), startCount+1)
 
+	// Make a whole table and make sure our app directory missing shows up.
+	// This could be done otherwise, but we'd have to go find the site in the
+	// array first.
 	table := platform.CreateAppTable()
 	for _, site := range apps["local"] {
-		platform.RenderAppRow(table, site)
+		desc, err := site.Describe()
+		if err != nil {
+			t.Fatalf("Failed to describe site %s: %v", site.GetName(), err)
+		}
+
+		platform.RenderAppRow(table, desc)
 	}
 	assert.Contains(table.String(), fmt.Sprintf("app directory missing: %s", testDir))
 

--- a/pkg/plugins/platform/local_test.go
+++ b/pkg/plugins/platform/local_test.go
@@ -640,6 +640,62 @@ func TestRouterNotRunning(t *testing.T) {
 	}
 }
 
+// TestListWithoutDir prevents regression where ddev list panics if one of the
+// sites found is missing a directory
+func TestListWithoutDir(t *testing.T) {
+	// Set up tests and give ourselves a working directory.
+	assert := asrt.New(t)
+	testcommon.ClearDockerEnv()
+
+	// startCount is the count of apps at the start of this adventure
+	apps := platform.GetApps()
+	startCount := len(apps["local"])
+
+	testDir := testcommon.CreateTmpDir("TestStartWithoutDdevConfig")
+
+	defer testcommon.Chdir(testDir)()
+
+	err := os.MkdirAll(testDir+"/sites/default", 0777)
+	assert.NoError(err)
+	err = os.Chdir(testDir)
+	assert.NoError(err)
+
+	config, err := ddevapp.NewConfig(testDir, "")
+	assert.NoError(err)
+	config.Name = "junk"
+	config.AppType = "drupal7"
+	err = config.Write()
+	assert.NoError(err)
+
+	// Do a start on the configured site.
+	app, err := platform.GetActiveApp("")
+	assert.NoError(err)
+	err = app.Start()
+	assert.NoError(err)
+
+	// Make sure we move out of the directory for Windows' sake
+	garbageDir := testcommon.CreateTmpDir("RestingHere")
+	defer testcommon.CleanupDir(garbageDir)
+
+	err = os.Chdir(garbageDir)
+	assert.NoError(err)
+
+	testcommon.CleanupDir(testDir)
+
+	apps = platform.GetApps()
+
+	assert.EqualValues(len(apps["local"]), startCount+1)
+
+	table := platform.CreateAppTable()
+	for _, site := range apps["local"] {
+		platform.RenderAppRow(table, site)
+	}
+	assert.Contains(table.String(), fmt.Sprintf("app directory missing: %s", testDir))
+
+	err = app.Down(true)
+	assert.NoError(err)
+}
+
 // constructContainerName builds a container name given the type (web/db/dba) and the app
 func constructContainerName(containerType string, app platform.App) (string, error) {
 	container, err := app.FindContainerByType(containerType)

